From 5a3daaaddbdc8ca2816bb00164a24a2b74a81b26 Mon Sep 17 00:00:00 2001
From: Pietro Abate <pietro.abate@pps.jussieu.fr>
Date: Thu, 19 Sep 2013 12:14:54 +0200
Subject: [PATCH 05/10] Fix upgrade bug in the check_request function

the upgrade request is not correctly encoded and passed to
the edos solver.

Conflicts:
	cudf
---
 algo/depsolver.ml | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/algo/depsolver.ml b/algo/depsolver.ml
index 869d153..6c5092a 100644
--- a/algo/depsolver.ml
+++ b/algo/depsolver.ml
@@ -337,7 +337,18 @@ let check_request ?cmd ?criteria ?(explain=false) (pre,universe,request) =
           else acc
         ) [] universe
       in
-      let l = request.Cudf.install @ request.Cudf.upgrade in
+      let il = request.Cudf.install in
+      let ul = 
+        List.filter_map (function (name,None) ->
+          match Cudf.get_installed universe name with
+          |[] -> Some((name,None))
+          |[p] -> Some(name,Some(`Geq,p.Cudf.version))
+          |pl ->
+              let p = List.hd(List.sort ~cmp:Cudf.(>%) pl) in
+              Some(name,Some(`Geq,p.Cudf.version))
+        ) request.Cudf.upgrade  
+      in
+      let l = il @ ul in
       debug "request consistency (keep %d) (install %d) (upgrade %d) (remove %d) (# %d)"
       (List.length k) (List.length request.Cudf.install) 
       (List.length request.Cudf.upgrade)
-- 
1.8.5.3

